# Meminimumkan
Akan dicari titik minimum dari 
f(x) = x1^2 + 5x2^2 + 25x3^2 - x1 - x2 - x3 


In [ ]:
import numpy as np



# Metode Algoritma Umum



In [ ]:
x=np.array([0,0,0]) #nilai awal
alpha=0.01  #coba kombinasi (0,0,0,0.1) , (0,0,0,0.11), (0,0,0,0.05), (1,1,1,0.1) , (1,1,1,0.01), (1,1,1,0.05)
nablaf=np.array([2*x[0]-1 , 10*x[1]-1 , 50*x[2]-1]) #menghitung nabla f(x) dengan x awal

print("x0= ",x)
print("nabla f(x0)= ",nablaf)
print("norm(nabla f(x0))= ",np.linalg.norm(nablaf))

error=0.00001
iter=1
maxiter=200


while  np.linalg.norm(nablaf)>error and iter<maxiter:
    x = np.array(x - (alpha*np.array(nablaf))) #memperbaharui x
    nablaf=np.array([2*x[0]-1, 10*x[1]-1, 50*x[2]-1]) #menghitung nabla f(x) yang baru dengan x baru
    if(iter%5==0):
        print("\n","x",iter,"= ",x)
        print("nabla f(x",iter,")= ",nablaf)
        print("norm(nabla f(xx",iter,")= ",np.linalg.norm(nablaf))
    iter=iter+1

print("\n","x",iter-1,"= ",x)
print("nabla f(x",iter-1,")= ",nablaf)
print("norm(nabla f(xx",iter-1,")= ",np.linalg.norm(nablaf))


x0=  [0 0 0]
nabla f(x0)=  [-1 -1 -1]
norm(nabla f(x0))=  1.7320508075688772

 x 5 =  [0.0480396 0.040951  0.019375 ]
nabla f(x 5 )=  [-0.9039208 -0.59049   -0.03125  ]
norm(nabla f(xx 5 )=  1.080151752990082

 x 10 =  [0.0914636  0.06513216 0.01998047]
nabla f(x 10 )=  [-0.81707281 -0.34867844 -0.00097656]
norm(nabla f(xx 10 )=  0.8883611765605136

 x 15 =  [0.13071545 0.07941089 0.01999939]
nabla f(x 15 )=  [-7.38569103e-01 -2.05891132e-01 -3.05175781e-05]
norm(nabla f(xx 15 )=  0.7667303819394246

 x 20 =  [0.16619601 0.08784233 0.01999998]
nabla f(x 20 )=  [-6.67607972e-01 -1.21576655e-01 -9.53674316e-07]
norm(nabla f(xx 20 )=  0.6785877149590134

 x 25 =  [0.19826764 0.09282102 0.02      ]
nabla f(x 25 )=  [-6.03464730e-01 -7.17897988e-02 -2.98023224e-08]
norm(nabla f(xx 25 )=  0.6077198822602713

 x 30 =  [0.22725784 0.09576088 0.02      ]
nabla f(x 30 )=  [-5.45484319e-01 -4.23911583e-02 -9.31322575e-10]
norm(nabla f(xx 30 )=  0.5471290094594101

 x 35 =  [0.25346269 0.09749684 

# Metode Steepest Descent



In [ ]:
#Metode Steepest Decent

x=np.array([0,0,0]) #nilai awal
nablaf=np.array([2*x[0]-1 , 10*x[1]-1 , 50*x[2]-1]) #menghitung nabla f(x) dengan x awal
nabla2f=np.array([[2,0,0],[0,10,0],[0,0,50]]) #kebetulan konstan, tidak bergantung pada x
alpha = (np.transpose(nablaf)@nablaf)/(np.transpose(nablaf)@nabla2f@nablaf)

print("x0= ",x)
print("nabla f(x0)= ",nablaf)
print("norm(nabla f(x0))= ",np.linalg.norm(nablaf))
print("alpha = ", alpha)


error=0.00001
iter=1
maxiter=100

while  np.linalg.norm(nablaf)>error and iter<maxiter:
  x=x-alpha*nablaf              #memperbaharui x
  print("\n","x",iter,"= ",x)
  nablaf=np.array([2*x[0]-1 , 10*x[1]-1 , 50*x[2]-1]) #menghitung nabla f(x) dengan x awal
  print("nabla f(x",iter,")= ",nablaf)
  print("norm(nabla f(x",iter,")= ",np.linalg.norm(nablaf))
  alpha = (np.transpose(nablaf)@nablaf)/(np.transpose(nablaf)@nabla2f@nablaf)
  print("alpha =" ,alpha)
  iter=iter+1

x0=  [0 0 0]
nabla f(x0)=  [-1 -1 -1]
norm(nabla f(x0))=  1.7320508075688772
alpha =  0.04838709677419355

 x 1 =  [0.0483871 0.0483871 0.0483871]
nabla f(x 1 )=  [-0.90322581 -0.51612903  1.41935484]
norm(nabla f(x 1 )=  1.7597653802562396
alpha = 0.029486366518706404

 x 2 =  [0.07501994 0.06360587 0.00653548]
nabla f(x 2 )=  [-0.84996011 -0.36394133 -0.67322601]
norm(nabla f(x 2 )=  1.1437301905217512
alpha = 0.05143783730671558

 x 3 =  [0.11874005 0.08232622 0.04116477]
nabla f(x 3 )=  [-0.76251989 -0.17673778  1.05823849]
norm(nabla f(x 3 )=  1.316260434723818
alpha = 0.030147583201209565

 x 4 =  [0.14172819 0.08765444 0.00926144]
nabla f(x 4 )=  [-0.71654363 -0.12345561 -0.53692816]
norm(nabla f(x 4 )=  0.9038627687453126
alpha = 0.052390308476319626

 x 5 =  [0.17926813 0.09412232 0.03739127]
nabla f(x 5 )=  [-0.64146374 -0.05877684  0.86956343]
norm(nabla f(x 5 )=  1.082160345074698
alpha = 0.030287995598546624

 x 6 =  [0.19869678 0.09590255 0.01105394]
nabla f(x 6 )=  [-0.6

In [1]:
#Metode Steepest Decent

x=np.array([0,0,0]) #nilai awal
nablaf=np.array([8*x[0] , 2*x[1] ]) #menghitung nabla f(x) dengan x awal
nabla2f=np.array([[8,0],[0,2]]) #kebetulan konstan, tidak bergantung pada x
alpha = (np.transpose(nablaf)@nablaf)/(np.transpose(nablaf)@nabla2f@nablaf)

print("x0= ",x)
print("nabla f(x0)= ",nablaf)
print("norm(nabla f(x0))= ",np.linalg.norm(nablaf))
print("alpha = ", alpha)


error=0.00001
iter=1
maxiter=100

while  np.linalg.norm(nablaf)>error and iter<maxiter:
  x=x-alpha*nablaf              #memperbaharui x
  print("\n","x",iter,"= ",x)
  nablaf=np.array([8*x[0] , 2*x[1] ]) #menghitung nabla f(x) dengan x awal
  print("nabla f(x",iter,")= ",nablaf)
  print("norm(nabla f(x",iter,")= ",np.linalg.norm(nablaf))
  alpha = (np.transpose(nablaf)@nablaf)/(np.transpose(nablaf)@nabla2f@nablaf)
  print("alpha =" ,alpha)
  iter=iter+1

NameError: ignored

# Metode Newton

In [ ]:
#Metode Newton

x=np.array([0,0,0]) #nilai awal
nablaf=np.array([2*x[0]-1 , 10*x[1]-1 , 50*x[2]-1]) #menghitung nabla f(x) dengan x awal
nabla2f=np.array([[2,0,0],[0,10,0],[0,0,50]]) #kebetulan konstan, tidak bergantung pada x

print("x0= ",x)
print("nabla f(x0)= ",nablaf)
print("norm(nabla f(x0))= ",np.linalg.norm(nablaf))

error=0.00001
iter=1
maxiter=100

while  np.linalg.norm(nablaf)>error and iter<maxiter:
  x=x-np.linalg.inv(nabla2f)@nablaf #memperbaharui x
  print("\n","x",iter,"= ",x)
  nablaf=np.array([2*x[0]-1 , 10*x[1]-1 , 50*x[2]-1]) #menghitung nabla f(x) dengan x awal
  nabla2f=np.array([[2,0,0],[0,10,0],[0,0,50]]) #kebetulan konstan, tidak bergantung pada x
  print("nabla f(x",iter,")= ",nablaf)
  print("norm(nabla f(x",iter,")= ",np.linalg.norm(nablaf))
  iter=iter+1

x0=  [0 0 0]
nabla f(x0)=  [-1 -1 -1]
norm(nabla f(x0))=  1.7320508075688772

 x 1 =  [0.5  0.1  0.02]
nabla f(x 1 )=  [0. 0. 0.]
norm(nabla f(x 1 )=  0.0


# Metode Conjugate Gradien

In [ ]:
# metode conjugate gradien

x=np.array([0,0,0]) #nilai awal
nablaf=np.array([2*x[0]-1 , 10*x[1]-1 , 50*x[2]-1]) #menghitung nabla f(x) dengan x awal
nabla2f=np.array([[2,0,0],[0,10,0],[0,0,50]])
d=-nablaf #menghitung d

print("x0= ",x)
print("norm(nabla f(x0))= ",np.linalg.norm(nablaf))


iter=1
maxiter=100
error=0.000001

while  np.linalg.norm(nablaf)>error and iter<maxiter:
  alpha=-d@nablaf/(d@nabla2f@d)
  x=x+alpha*d
  print("x",iter,"= ", x)
  nablaf=np.array([2*x[0]-1 , 10*x[1]-1 , 50*x[2]-1])
  print("norm(nabla f(x",iter,"))= ",np.linalg.norm(nablaf)) 
  d=-nablaf+( nabla2f@nablaf@d/(d@nabla2f@d)  )*d
  iter=iter+1

x0=  [0 0 0]
norm(nabla f(x0))=  1.7320508075688772
x 1 =  [0.0483871 0.0483871 0.0483871]
norm(nabla f(x 1 ))=  1.7597653802562396
x 2 =  [0.20223325 0.17146402 0.01761787]
norm(nabla f(x 2 ))=  0.9378470916937143
x 3 =  [0.5  0.1  0.02]
norm(nabla f(x 3 ))=  3.614624287906422e-15


# Nilai Minimum f(x)

In [ ]:
# Nilai Minimum f(x)
x = np.array([0.5, 0.1, 0.02])
def f(x1,x2,x3):
    return x1**2 + 5*x2**2 + 25*x3**2 - x1 - x2 - x3 
z = f(0.5, 0.1, 0.02)
print("Nilai Minimum f(x) = ",z)

Nilai Minimum f(x) =  -0.31000000000000005


# Kesimpulan
Dari hasil metode-metode di atas, diperoleh titik minimum f(x) ada di x = [0.5 , 0.1, 0.02] dengan nilai minimum f(x) adalah -0.31